### PR TITLE
Upcoming blog post on filtering behaviour

### PR DESCRIPTION
**Pull request by jobergum**
TODO: 

- Another pass, the summary is pretty weak 
- Add links to reference doc with new parameters 
- Mention explicit Vespa version these parameters were introduced. 

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+evaluated. In the the case of *pre-filtering*, two additional hits were exposed, but they 
+are more distant neighbors. 
+</em>
+
+### Filtering strategies and serving performance
+From a performance and serving cost perspective, one can on a high level summarize the filtering strategies:
+
+* Pure approximate nearest neighbor search without filters is the cheapest.  
+* *Post-filtering* is less resource-intensive than pre-filtering for the same number of `targetHits`. 
+
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:257]:
```suggestion
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+*Pre-filtering* uses more resources than *post-filtering* as it needs to execute the
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:258]:
```suggestion
* *Pre-filtering* uses more resources than *post-filtering* as it needs to execute the 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+result (few hits) causes the greedy graph search to traverse many nodes before finding the `targetHits` 
+which satisfies the document ID list from the pre-filter execution.  
+Due to this, there is a sweet spot or threshold where an exact search with filters
+has a lower computational cost than an approximate search using the document ID list from the pre-filter execution. 
+The actual threshold value depends on vector dimensionality, `HNSW` graph properties, 
+and the number of vectors indexed in the Vespa instance. 
+</em>
+
+## Controlling the filtering behavior with approximate nearest neighbor search 
+Vespa exposes two parameters that control the query time filtering strategy.
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:273]:
```suggestion
Vespa exposes two parameters that control the query-time filtering strategy. 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+These parameters were introduced in Vespa 7.580.54, 
+and can be configured in the [rank profile](https://docs.vespa.ai/en/reference/schema-reference.html#rank-profile), 
+defined in the schema, or set using the [query API](https://docs.vespa.ai/en/reference/query-api-reference.html) on
+a per-query request basis. The query api parameters are:
+
+* **ranking.matching.postFilterThreshold** - default 1.0 
+* **ranking.matching.approximateThreshold** - default 0.05
+
+<img src="/assets/2022-05-09-constrained-approximate-nearest-neighbor-search/flowchart-small.png"/>
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:290]:
```suggestion
<img src="/assets/2022-05-09-constrained-approximate-nearest-neighbor-search/flowchart.png"/>
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+
+<pre>
+rank-profile hybrid-filtering {
+    post-filter-threshold: 0.75
+    approximate-threshold: 0.05
+}
+</pre>
+This parameter combination will trigger *post-filtering* for relaxed filters, 
+estimated to match more than 75% of the documents. 
+Moderate filters (between 5% and 75%) are evaluated using `pre-filtering`
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:357]:
```suggestion
Moderate filters (between 5% and 75%) are evaluated using *pre-filtering* 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    approximate-threshold: 0.05
+}
+</pre>
+This parameter combination will trigger *post-filtering* for relaxed filters, 
+estimated to match more than 75% of the documents. 
+Moderate filters (between 5% and 75%) are evaluated using `pre-filtering` 
+and restrictive filters (&lt; 5%) are evaluated using exact search.  
+
+## Summary
+Constraining search for nearest neighbors using query time constraints is mission-critical for
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:361]:
```suggestion
Constraining search for nearest neighbors using query-time constraints is mission-critical for 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+Constraining search for nearest neighbors using query time constraints is mission-critical for 
+real-world applications using AI-powered vector representations. By introducing these new native parameters for controlling vector search filtering behavior,  
+Vespa further fortifies its position in the industry as the leading open-source serving technology for vector search applications. 
+
+To try out these parameters in practice, see the [practical nearest neighbor search guide](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html) and
+especially the section on 
+[controlling filter behavior](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html#controlling-filter-behavior). 
+
+See also [Vespa sample applications](https://github.com/vespa-engine/sample-apps) built using Vespa's approximate 
+nearest neighbor search support to build:
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:370]:
```suggestion
nearest neighbor search:
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+
+- [State-of-the-art text ranking](https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md) 
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
+The application also demonstrates multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html) 
+serving pipelines. Furthermore, the app demonstrates how embedding models, which maps the text data to vector representation, can be
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:375]:
```suggestion
serving pipelines. Furthermore, the application demonstrates how embedding models, which maps the text data to vector representation, can be 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
+The application also demonstrates multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html) 
+serving pipelines. Furthermore, the app demonstrates how embedding models, which maps the text data to vector representation, can be 
+deployed to Vespa for [run-time inference](https://blog.vespa.ai/stateless-model-evaluation/) during both document and query processing.
+
+- [State-of-the-art image search](https://github.com/vespa-engine/sample-apps/tree/master/text-image-search) Using AI-powered multi-modal vector representations
+to retrieve images for a text query. 
+
+- [State-of-the art open-domain question answering](https://github.com/vespa-engine/sample-apps/tree/master/dense-passage-retrieval-with-ann) Using AI-powered vector representations
+to retrieve passages from Wikipedia which are feed into a NLP reader model which identifies the answer. End-to-end represented using Vespa.
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:382]:
```suggestion
to retrieve passages from Wikipedia which are fed into a NLP reader model which identifies the answer. End-to-end represented using Vespa.
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+All these are examples of applications built using AI-powered vector representations, and where real-world deployments 
+need query time constrained nearest neighbor search.
```
Comment (geirst) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:385]:
```suggestion
need query-time constrained nearest neighbor search. 
```

**Review by geirst — COMMENTED** [review]
Very good 👍

**Comment by jobergum** [comment]
Would you mind giving it a pass with new eyes @kkraune - no hurry, we aim to publish this on Monday.

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+To try out these parameters in practice, see the [practical nearest neighbor search guide](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html) and
+especially the section on 
+[controlling filter behavior](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html#controlling-filter-behavior). 
+
+See also [Vespa sample applications](https://github.com/vespa-engine/sample-apps) built using Vespa's approximate 
+nearest neighbor search:
+
+- [State-of-the-art text ranking](https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md) 
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval.
```
Comment (kkraune) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:372]:
```suggestion
Vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+To try out these parameters in practice, see the [practical nearest neighbor search guide](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html) and
+especially the section on 
+[controlling filter behavior](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html#controlling-filter-behavior). 
+
+See also [Vespa sample applications](https://github.com/vespa-engine/sample-apps) built using Vespa's approximate 
+nearest neighbor search:
+
+- [State-of-the-art text ranking](https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md) 
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
+The application also demonstrates multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html)
```
Comment (kkraune) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:373]:
```suggestion
The application has multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html) 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+especially the section on 
+[controlling filter behavior](https://docs.vespa.ai/en/nearest-neighbor-search-guide.html#controlling-filter-behavior). 
+
+See also [Vespa sample applications](https://github.com/vespa-engine/sample-apps) built using Vespa's approximate 
+nearest neighbor search:
+
+- [State-of-the-art text ranking](https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md) 
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
+The application also demonstrates multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html) 
+serving pipelines. Furthermore, the application demonstrates how embedding models, which maps the text data to vector representation, can be
```
Comment (kkraune) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:374]:
```suggestion
serving pipelines. Furthermore, the application shows how embedding models, which maps the text data to vector representation, can be 
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+Demonstrating using vector search with AI-powered representations built on NLP Transformer models for candidate retrieval. 
+The application also demonstrates multi-vector representations for re-ranking, using Vespa's [phased retrieval and ranking](https://docs.vespa.ai/en/phased-ranking.html) 
+serving pipelines. Furthermore, the application demonstrates how embedding models, which maps the text data to vector representation, can be 
+deployed to Vespa for [run-time inference](https://blog.vespa.ai/stateless-model-evaluation/) during both document and query processing.
+
+- [State-of-the-art image search](https://github.com/vespa-engine/sample-apps/tree/master/text-image-search) Using AI-powered multi-modal vector representations
+to retrieve images for a text query. 
+
+- [State-of-the art open-domain question answering](https://github.com/vespa-engine/sample-apps/tree/master/dense-passage-retrieval-with-ann) Using AI-powered vector representations
+to retrieve passages from Wikipedia which are fed into a NLP reader model which identifies the answer. End-to-end represented using Vespa.
```
Comment (kkraune) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:381]:
```suggestion
to retrieve passages from Wikipedia, which are fed into a NLP reader model which identifies the answer. End-to-end represented using Vespa.
```

File: _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+From a performance and serving cost perspective, one can on a high level summarize the filtering strategies:
+
+* Pure approximate nearest neighbor search without filters is the cheapest.  
+* *Post-filtering* is less resource-intensive than pre-filtering for the same number of `targetHits`. 
+* *Pre-filtering* uses more resources than *post-filtering* as it needs to execute the 
+filter and then search the `HNSW` graph, constrained by the document ID list produced by the pre-filter execution. 
+
+
+<img src="/assets/2022-05-09-constrained-approximate-nearest-neighbor-search/pre-filter-approximate-threshold.png"/>
+<em>**Figure 4** With the pre-filtering strategy, searching the `HNSW` graph with a restrictive pre-filter
```
Comment (kkraune) [inline _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:262]: the image caption is very long, consider moving some to regular text

Comment (jobergum) [reply _posts/2022-05-09-constrained-approximate-nearest-neighbor-search.md:262]: Thanks

**Review by geirst — APPROVED** [review]
👍  This looks very good. Some minor typing comments.

**Pull request merged**